### PR TITLE
Mjv/preferences

### PR DESCRIPTION
**Pull request by marcJV**
Added a global user preferences to save your last selected sprite, name, and host

Added logic around empty names (will default to the skin's name, and the default host)

Added in controller rebinding logic, currently only reading from file

### CODE REVIEW

**Comment by marcJV** [comment]
Need to add ability to update controls, everything else works right now

**Comment by Dobbelsteen** [comment]
Cool! Didn't go that into depth in the code, but it looks pretty nice.

My one nitpick is the `set_pref `and `get_pref` function names, I'd just name them `get()` and `set()` to keep it simple and easy to remember
You know you're accessing those function from the user_preferences Node, so it's logical you want to get or set pa certain preference.

Other than that, props to you :)

File: engine/lobby.gd [hunk 1]
```diff
@@ -64,8 +65,25 @@ func _set_status(text,isok):
...
+	
+	user_preferences.set_pref("host_address", ip)
+	
+	return ip
+
 func _on_host_pressed():
-	network.my_player_data.name = $characterselect/name.text
+	network.my_player_data.name = $characterselect.player_name
+	
+	var ip = check_host_address(get_node("panel/address").get_text())
```
Comment (Dobbelsteen) [inline engine/lobby.gd]: Don't think we need to verify the IP when we're hosting a game

File: engine/user_preferences.gd [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+# If we remove a key, currently that data is lost
+# When needed, we can start migrating using the schema numbers to do more complicated migrations
+func _migrate_save(loaded_save):
+	for key in _user_prefs.keys():
+		if loaded_save.has(key):
+			_user_prefs[key] = loaded_save[key]
+	
+	_save_to_preferences()
+	
+func get_pref(key: String):
```
Comment (Dobbelsteen) [inline engine/user_preferences.gd]: I'd rename it to a simple get()

File: engine/user_preferences.gd [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+	if _user_prefs.has(key):
+		return _user_prefs[key]
+		
+	printerr( "Key `%s` not found! " % key)
+	return null
+	
+func set_pref(key: String, value):
```
Comment (Dobbelsteen) [inline engine/user_preferences.gd]: And set()

File: engine/user_preferences.gd (hunk 2 above)
Comment (marcJV) [reply engine/user_preferences.gd]: get would override behavior in this case.
I agree that get_pref isnt ideal, will think about this and update it

Comment (Dobbelsteen) [reply engine/user_preferences.gd]: eh, good enough then!

**Pull request merged**